services/src/main/java/org/adamalang/services/social/Discord.java

```diff
@@ -22,46 +22,227 @@
+
+    public DiscordHttpResponder(Logger logger, RequestResponseMonitor.RequestResponseMonitorInstance monitor, Callback<String> callback) {
+      this.logger = logger;
+      this.monitor = monitor;
+      this.callback = callback;
+      this.invokeSuccess = false;
+      this.emissionPossible = true;
+      this.logBody = false;
+    }
+    @Override
+    public void start(SimpleHttpResponseHeader header) {
+      if (emissionPossible) {
+        if (200 <= header.status && header.status <= 204) {
+          invokeSuccess = true;
+        } else {
+          emissionPossible = false;
+          logBody = HttpError.convert(header, logger,  ErrorCodes.WEB_STRING_CALLBACK_NOT_200, monitor, callback);
+        }
+      }
+    }
+
+    @Override
+    public void bodyStart(long size) {
+      if (size > 0) {
+        this.memory = new ByteArrayOutputStream((int) size);
+      }
+    }
+
+    @Override
+    public void bodyFragment(byte[] chunk, int offset, int len) {
+      if (memory == null) { // unknown size due to an error
+        this.memory = new ByteArrayOutputStream();
+      }
+      memory.write(chunk, offset, len);
+    }
+
+    @Override
+    public void bodyEnd() {
+      if (invokeSuccess && emissionPossible) {
+        String callbackVal = "";
+        if (memory != null) {
+          callbackVal = new String(memory.toByteArray(), StandardCharsets.UTF_8);
+        }
+        callback.success(callbackVal);
+        monitor.success();
+      }
+      if (logBody) {
+        if (memory != null) {
+          System.out.println(new String(memory.toByteArray(), StandardCharsets.UTF_8));
+        }
+        logger.error("failed body: {}", new String(memory.toByteArray(), StandardCharsets.UTF_8));
+      }
+    }
+
+    @Override
+    public void failure(ErrorCodeException ex) {
+      if (emissionPossible) {
+        emissionPossible = false;
+        logger.error("string-callback-failure:", ex);
+        monitor.failure(ex.code);
+        callback.failure(ex);
+      }
+    }
+  }
+  private void apiCall(String method, String[] uri, String[] requiredParams, TreeMap<String, String> headers, ObjectNode requestNode, String body, Callback<String> callback) {
+    for (String params : requiredParams) {
+      String hasParam = params;
+      if (params.startsWith("_")) {
+        hasParam = params.substring(1);
+      }
+      if (!requestNode.has(hasParam)) {
+        callback.failure(new ErrorCodeException(888890, "Method is missing parameter: " + hasParam));
+        return;
+      }
+    }
+    StringBuilder uriBuilder = new StringBuilder();
+    uriBuilder.append("https://discord.com/api");
+    int pos = 0;
+    while (pos < (requiredParams.length)) {
+      if (requiredParams[pos].startsWith("_")) {
+        break;
+      }
+      pos++;
+      uriBuilder.append(uri[pos - 1]).append("/" + requestNode.get(requiredParams[pos - 1]).textValue());
+    }
+    for (int i = pos; i < uri.length; i++) {
+      uriBuilder.append(uri[i]);
+    }
```

Comment:
This feels almost worth of a separate class with unit testing, and a comment of what _ means in this would be fantastic

Reply:
I decided not to create a separate class, because I did not know what the class would be named, and if I did have a class, it would only have one method. 

I am still willing to create the class though, because unit testing would be good.

Reply:
Naming is hard, and I'd probably call it... DiscordRequestBuilder